app/src/main/java/de/jakobclass/transittracker/MapActivity.kt

```diff
@@ -1,43 +1,102 @@
 
+    override fun onStop() {
+        googleApiClient?.disconnect()
+        super.onStop()
+    }
 
-    /**
-     * Manipulates the map once available.
-     * This callback is triggered when the map is ready to be used.
-     * This is where we can add markers or lines, add listeners or move the camera. In this case,
-     * we just add a marker near Sydney, Australia.
-     * If Google Play services is not installed on the device, the user will be prompted to install
-     * it inside the SupportMapFragment. This method will only be triggered once the user has
-     * installed Google Play services and returned to the app.
-     */
     override fun onMapReady(googleMap: GoogleMap) {
         map = googleMap
 
-        // Add a marker in Sydney and move the camera
-        val sydney = LatLng(-34.0, 151.0)
-        map!!.addMarker(MarkerOptions().position(sydney).title("Marker in Sydney"))
-        map!!.moveCamera(CameraUpdateFactory.newLatLng(sydney))
+        val berlinCityCenter = LatLng(52.520048, 13.404773)
+        map!!.moveCamera(CameraUpdateFactory.newLatLngZoom(berlinCityCenter, 16.0f))
+        initLocationPermission()
+    }
+
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
+        if (requestCode == REQUEST_CODE_ACCESS_FINE_LOCATION && permissions.isNotEmpty()
+                && permissions[0] == Manifest.permission.ACCESS_FINE_LOCATION
+                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            enableMyLocation()
+        }
+    }
+
+    private fun initLocationPermission() {
+        if (ContextCompat.checkSelfPermission(this, REQUIRED_LOCATION_PREMISSION)
+                == PackageManager.PERMISSION_GRANTED) {
+            enableMyLocation()
+        } else if (ActivityCompat.shouldShowRequestPermissionRationale(this, REQUIRED_LOCATION_PREMISSION)) {
+            AlertDialog.Builder(this).setMessage(R.string.location_usage_description)
+                    .setNegativeButton(R.string.cancel, null)
+                    .setPositiveButton(R.string.ok) { dialog, index -> requestLocationPermission() }
+                    .show()
+        } else {
+            requestLocationPermission()
+        }
+    }
+
+    private fun requestLocationPermission() {
+        ActivityCompat.requestPermissions(this, arrayOf(REQUIRED_LOCATION_PREMISSION), REQUEST_CODE_ACCESS_FINE_LOCATION)
     }
+
+    private fun enableMyLocation() {
+        map?.isMyLocationEnabled = true
+        if (googleApiClient == null) {
+            googleApiClient = GoogleApiClient.Builder(this)
+                    .addApi(LocationServices.API)
+                    .addConnectionCallbacks(this)
+                    .build()
+        }
+        googleApiClient?.connect()
+    }
+
+    override fun onConnected(connectionHint: Bundle?) {
+        val lastLocation = LocationServices.FusedLocationApi.getLastLocation(googleApiClient)
+        lastLocation?.let {
```

Comment:
Hmmm, interesting, I found this one: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/let.html https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/src/kotlin/util/Standard.kt#L55, there's no check for `null`.

Ah, wait...it's complining into `data.let {} if data != null` so it makes sense now.

But then you can use `it` inside the block instead of `lastLocation`